<!DOCTYPE html>
<html lang="en">

<head>
  <meta charset="UTF-8">
  <meta name="viewport" content="width=device-width, initial-scale=1.0">
  <meta http-equiv="X-UA-Compatible" content="ie=edge">
  <title>LocusFocus: A Colocalization Tool to Prioritize Genes and Tissues from GWAS and eQTL Integration</title>
  <link rel="shortcut icon" href="#" />
   <!-- Font Awesome -->
   <link rel="stylesheet" href="https://use.fontawesome.com/releases/v5.7.0/css/all.css">
   <!-- Bootstrap core CSS -->
   <link rel="stylesheet" href="https://stackpath.bootstrapcdn.com/bootstrap/4.3.1/css/bootstrap.min.css" integrity="sha384-ggOyR0iXCbMQv3Xipma34MD+dH/1fQ784/j6cY/iJTQUOhcWr7x9JvoRxT2MZw1T" crossorigin="anonymous">
   <!-- Material Design Bootstrap -->
   <link href="https://cdnjs.cloudflare.com/ajax/libs/mdbootstrap/4.7.4/css/mdb.min.css" rel="stylesheet">
   <!-- DataTables -->
   <link rel="stylesheet" href="{{ url_for('static', filename='css/jquery.dataTables.min.css') }}"/>
   <link rel="stylesheet" href="{{ url_for('static', filename='DataTables/datatables.min.css') }}"/>
   <link rel="stylesheet" type="text/css" href="{{ url_for('static', filename='css/style.css') }}"/>
   <link rel="stylesheet" type="text/css" href="{{ url_for('static', filename='css/d3Style.css') }}"/>
</head>

<body>
    <div class="container">
    <div class="row" id="myjumbotron"></div>
    <div class="row" id="navigation-bar"></div>
    <div class="row">
        <div class="col-md-12 d-flex justify-content-center align-items-center">
            <div class="card border-primary mb-9" style="max-width: 60rem;">
                <div class="card-header">Session ID</div>
                <div class="card-body">
                    <h4 class="card-title" id="sessionid"></h4>
                    <p class="card-text">Save the above string for your records to load or share your plot.</p>
                    <p class="card-text">Plots older than 7 days are deleted.</p>
                </div>
            </div>
        </div>
    </div>
    <hr/>
    <div class="row">
        <div class="col-md-12">
            <h3>Colocalization Plot</h3>
            <select id="selGene" onchange="optionChanged(this.value)"></select>
            <div id="plot_message" class="isa_info"></div>
            
            <div class="row">
                <div class="col-md-6">
                    <input type="checkbox" id="pval-filter" name="pval-filter" onclick="plot_fullgwas(this)" checked><label for="pval-filter" style="margin: 0px; padding: 10px;  height:42px"
                    data-toggle="tooltip" data-html="true" title="<p>If unchecked, all GWAS p-values above 0.1 will be shown (slower to load and plot).</p>""> Uncheck to draw all GWAS points below -log<sub>10</sub>P &lt; 1 (slow)
                </div>
            </div>
            <div id="plot"></div>
            <div class="row">
                <div class="col-md-3">
                    <input class="form-control" id="colocPlotWidth" name="colocPlotWidth" type="text" placeholder="Width (e.g. 1080)">
                </div>
                <div class="col-md-3">
                    <input class="form-control" id="colocPlotHeight" name="colocPlotHeight" type="text" placeholder="Height (e.g. 1080)">
                </div>
                <div class="col-md-3">
                    <input class="form-control" id="eqtlWindow" name="eqtlWindow" type="text" placeholder="Lines window size (e.g. 300)" data-toggle="tooltip" title="This determines the window size for the minP line traces; use -1 for default">
                </div>
                <div class="col-md-3">
                    <input class="form-control" id="pctOneChar" name="pctOneChar" type="text" placeholder="One character space (e.g. 0.020)" data-toggle="tooltip" title="This is the fraction occupied by one character for the gene names; change this if gene names are not appearing">
                </div>
            </div>
            <br>
            <div class="row">
                <div class="col-md-3">
                    <input class="form-control" id="colocPlotFontSize" name="colocPlotFontSize" type="text" placeholder="Global font size (e.g. 14)">
                </div>
                <div class="col-md-3">
                    <input class="form-control" id="legendOffset" name="legendOffset" type="text" placeholder="Legend offset (e.g. 0.1)" data-toggle="tooltip" title="If the legend is overlapping the plot, you may want to increase the offset">
                </div>
            </div>
            <div class="row">
                <div class="col-md-3">
                    <input id="colocPlot-redraw-btn" type="submit" class="btn btn-default" value="Update">
                </div>
                <div class="col-md-3">
                    <input id="colocPlot-svg-btn" type="submit" class="btn btn-default" value="Generate SVG">
                </div>
            </div>
            <p>
            <div class="alert alert-dismissible alert-info">
                <button type="button" class="close" data-dismiss="alert">&times;</button>
                <strong>Tissues selected but not drawn are due to unavailable eQTL data, likely due to little or no expression in that tissue</strong>
            </div>
        </div>
    </div>
    <hr/>
    <div class="row">
        <div class="col-md-12">
            <h3>Heatmap of Simple Sum Colocalization P-values for GTEx Tissues Selected</h3>
            <div id="heatmap"></div>
            <div class="gtex_heatmap_message"></div>
            <div class="row">
                <div class="col-md-3">
                    <!-- <label for="heatmapwidth" id="heatmap-width">Width</label> -->
                    <input class="form-control" id="heatmapWidth" name="heatmapWidth" type="text" placeholder="Width (e.g. 1080)">
                </div>
                <div class="col-md-3">
                    <!-- <label for="heatmapheight" id="heatmap-height">Height</label> -->
                    <input class="form-control" id="heatmapHeight" name="heatmapHeight" type="text" placeholder="Height (e.g. 1080)">
                </div>
                <div class="col-md-3">
                        <input class="form-control" id="heatmapFontsize" name="heatmapFontsize" type="text" placeholder="Global font size (e.g. 14)">
                </div>
            </div>
            <div class="row">
                <div class="col-md-3">
                    <input id="heatmap-redraw-btn" type="submit" class="btn btn-default" value="Update">
                </div>
                <div class="col-md-3">
                    <input id="heatmap-svg-btn" type="submit" class="btn btn-default" value="Generate SVG">
                </div>
            </div>
            <p>
            <div class="alert alert-dismissible alert-info">
                <button type="button" class="close" data-dismiss="alert">&times;</button>
                <p><strong>Gray cells (with negative -log10 Simple Sum p-values) correspond to gene-tissue pairs with no Simple Sum (SS) P-value.<br>
                    The exact reason for the absence of a SS P-value can be gleaned in the table below </strong></p>
                <p><strong>Please note that we leave up to the user to determine the threshold of significance among the datasets that passed the first-stage significance test.
                    For example, if a user selected 3 tissues and 4 genes for testing, and 3 other secondary datasets (a total of 3 &times; 4 + 3 = 15 tests) and among these,
                    6 datasets were tested for colocalization, then one would conservatively choose to consider a Bonferroni-corrected p-value threshold of 
                    0.05 / 6 = 8.3 &times; 10<sup>-3</sup>. Thus, Simple Sum colocalization tests above this threshold would be considered as significant.</strong>
                </p>
            </div>
        </div>
    </div>
    <hr/>
    <div class="row">
        <div class="col-md-12">
            <div id="table-area">
                <h3>Simple Sum -log10 P-values Table for GTEx Tissues Selected</h3>
                <div class="gtex_heatmap_message"></div>

                <div class="row">
                    <div class="col-md-12"><input id="transpose-button" type="submit" name="transpose-button" class="btn btn-default" value="Transpose" onclick="transposeTable();"/></div>
                </div>

                <table id="variants-table" class="table table-striped table-bordered table-condensed table-sm sortable" cellspacing="0" width="90%">
                    <thead>
                        <tr>
                            <th class="th-sm" id="GeneOrTissue"></th>
                        </tr>
                    </thead>
                    <tbody></tbody>
                </table>
            </div> <!-- table-area closing div -->
            <div class="alert alert-dismissible alert-info">
                <button type="button" class="close" data-dismiss="alert">&times;</button>
                <p><strong>-1 values correspond to gene-tissue pairs with no eQTL data (likely due to little or no expression)</strong></p>
                <p><strong>-2 values correspond to gene-tissue pairs that did not pass the Bonferroni-corrected first stage testing for signficance among the secondary datasets chosen</strong></p>
                <p><strong>-3 values correspond to gene-tissue pairs where the Simple Sum P-value computation failed, likely due to insufficient SNPs</strong></p>
                <p><strong>Please note that we leave up to the user to determine the threshold of significance among the datasets that passed the first-stage significance test.
                    For example, if a user selected 3 tissues and 4 genes for testing, and 3 other secondary datasets (a total of 3 &times; 4 + 3 = 15 tests) and among these,
                    6 datasets were tested for colocalization, then one would conservatively choose to consider a Bonferroni-corrected p-value threshold of 
                    0.05 / 6 = 8.3 &times; 10<sup>-3</sup>. Thus, Simple Sum colocalization tests above this threshold would be considered as significant.</strong>
                </p>
            </div>
        </div> <!-- col-md-12 closing div -->
    </div> <!-- row closing div -->
    <hr/>

    <hr/>
    <div class="row">
        <div class="col-md-12">
            <div id="table-area">
                <h3>Simple Sum -log10 P-values Table for Secondary Datasets Uploaded</h3>
                <div id="secondary_datasets_message"></div>
                <!--
                <div class="row">
                    <div class="col-md-12"><input id="transpose-button" type="submit" ame="transpose-button" class="btn btn-default" value="Transpose" onclick="transposeTable();"/></div>
                </div>
                -->
                <table id="secondary-table" class="table table-striped table-bordered table-condensed table-sm sortable" cellspacing="0" width="90%">
                    <thead>
                        <tr>
                            <th class="th-sm" id="dataset-desc"></th>
                        </tr>
                    </thead>
                    <tbody></tbody>
                </table>
            </div> <!-- table-area closing div -->
            <div class="alert alert-dismissible alert-info">
                <button type="button" class="close" data-dismiss="alert">&times;</button>
                <p><strong>-1 values correspond to gene-tissue pairs with no eQTL data (likely due to little or no expression)</strong></p>
                <p><strong>-2 values correspond to gene-tissue pairs that did not pass the Bonferroni-corrected first stage testing for signficance among the secondary datasets chosen</strong></p>
                <p><strong>-3 values correspond to gene-tissue pairs where the Simple Sum P-value computation failed, likely due to insufficient SNPs</strong></p>
                <p><strong>Please note that we leave up to the user to determine the threshold of significance among the datasets that passed the first-stage significance test.
                    For example, if a user selected 3 tissues and 4 genes for testing, and 3 other secondary datasets (a total of 3 &times; 4 + 3 = 15 tests) and among these,
                    6 datasets were tested for colocalization, then one would conservatively choose to consider a Bonferroni-corrected p-value threshold of 
                    0.05 / 6 = 8.3 &times; 10<sup>-3</sup>. Thus, Simple Sum colocalization tests above this threshold would be considered as significant.</strong>
                </p>
            </div>
        </div> <!-- col-md-12 closing div -->
    </div> <!-- row closing div -->
    <hr/>


    <hr/>
    <div class="row">
        <div class="col-md-12">
            <div id="coloc2-table-area">
                <h3>COLOC2 Posterior Probability Results Table</h3>
                <div id="coloc2-message"></div>
                <table id="coloc2-table" class="table table-striped table-bordered table-condensed table-sm sortable" cellspacing="0" width="90%">
                </table>
            </div>
        </div>
    </div>


    <hr/>
    <div class="row">
        <div class="col-md-4">
        <input id="download-btn" type="button" class="btn btn-default" value="Download Session Data" onclick="window.location.href = '/downloaddata/{{ sessionid }}';">
        </div>
    </div>

    </div> <!-- container closing div -->

    <!-- JS -->
  <script src="https://cdnjs.cloudflare.com/ajax/libs/d3/5.9.1/d3.js"></script>

  <!-- Filesaver -->
  <script type="module" src="{{ url_for('static', filename='js/file-saver/src/FileSaver.js') }}"></script>
  
  <!-- Plotly -->
  <script src="https://cdn.plot.ly/plotly-latest.min.js"></script>
  <script type="text/javascript" src="{{ url_for('static', filename='js/plot.js') }}"></script>
  <script type="text/javascript" src="{{ url_for('static', filename='js/plot_heatmap.js') }}"></script>
  <script type="text/javascript" src="{{ url_for('static', filename='js/heatmap_table.js') }}"></script>
  <script type="text/javascript" src="{{ url_for('static', filename='js/coloc2table.js') }}"></script>
  <script type="text/javascript">
    var thefile="{{ sessionfile }}";
    var genesfile="{{ genesfile }}";
    var SSPvalues_file="{{ SSPvalues_file }}";
    var coloc2_file="{{ coloc2_file }}";
    var sessionid="{{ sessionid }}";
    var transpose=false;
    function init() {
        // d3.select('#plot_message').append("p").attr("class","text-warning").text(`Please wait while loading colocalization plot`);
        d3.select('#plot_message').append("i").classed('fa',true).classed('fa-info-circle', true).text(`Please wait while loading colocalization plot`);
        d3.select("#sessionid").append('a').attr('href',`/session_id/${sessionid}`).text(sessionid);
        d3.json("{{ url_for('static', filename=sessionfile) }}").then(response => {
            d3.json("{{ url_for('static', filename=genesfile) }}").then(genesResponse => {
                d3.json("{{ url_for('static', filename=SSPvalues_file) }}").then(SSResponse => {
                    d3.json("{{ url_for('static', filename=coloc2_file) }}").then(coloc2Response => {
                        var data = response;
                        var genesdata = genesResponse;
                        populate_genes(genesdata, data);
                        if(SSResponse['Tissues'].length === 0) {
                            d3.selectAll('.gtex_heatmap_message').append("p").attr("class","text-warning").text(`No GTEx tissues were selected`);
                        }
                        if(SSResponse['Tissues'].length>0) {
                            plot_heatmap(SSResponse.Genes, SSResponse.Tissues, SSResponse.SSPvalues);
                            buildTable(SSResponse.Genes, SSResponse.Tissues, SSResponse.SSPvalues);
                        }
                        if(SSResponse['Secondary_dataset_titles'].length === 0) {
                            d3.select('#secondary_datasets_message').append("p").attr("class","text-warning").text(`No secondary datasets uploaded`);
                        }
                        if(SSResponse['Secondary_dataset_titles'].length>0) {
                            list_secondary_SSPvalues(SSResponse.Secondary_dataset_titles, SSResponse.SSPvalues_secondary);
                        }
                        if(coloc2Response['ProbeID'].length === 0) {
                            d3.select("#coloc2-message").append("p").attr("class","text-warning").text(`Insufficient data for COLOC2 calculations`);
                        }
                        if(coloc2Response['ProbeID'].length>0) {
                            buildColoc2Table(coloc2Response);
                        }
                        plot_gwas(data, genesdata);
                        d3.select('#plot_message').text("");

                        // initialize popover and tooltip
                        $(function () {
                            $('[data-toggle="popover"]').popover()
                        });
                        $(document).ready(function(){
                            $('[data-toggle="tooltip"]').tooltip({
                                delay: { "show": 500, "hide": 100 }
                            });
                        });
                    });
                });
            });
        });
    }

    function populate_genes(genesdata, response) {
        var selector = d3.select("#selGene");
        var genenames = genesdata.map(gene => gene['name']);
        var geneindex = 0;
        Object.keys(genenames).forEach((gene, i) => {
            if(genenames[i] === response['gene']) {
                geneindex = i;
            }
            selector
                .append("option")
                .text(genenames[i])
                .property('value', genenames[i]);
            
        });
        selector.property("selectedIndex", geneindex)
    }

    function optionChanged(newgene) {
        var newurl=`/update/${sessionid}/${newgene}`;
        d3.select('#plot').text("");
        // d3.select('#plot_message').append("p").attr("class","text-warning").text(`Please wait while re-drawing eQTLs for gene ${newgene}`);
        d3.select('#plot_message').append("i").classed('fa',true).classed('fa-info-circle', true).text(`Please wait while re-drawing eQTLs for gene ${newgene}`);
        // d3.json("{{ url_for('update_colocalizing_gene', session_id=sessionid, newgene=newgene) }}").then(response => {
        d3.json(newurl).then(newresponse => {
            d3.json("{{ url_for('static', filename=genesfile) }}").then(genesResponse => {
                var newdata = newresponse;
                var genesdata = genesResponse;
                plot_gwas(newdata, genesdata);
                d3.select('#plot_message').text("")
                // plot_heatmap(SSResponse.Genes, SSResponse.Tissues, SSResponse.SSPvalues);
                // buildTable(SSResponse.Genes, SSResponse.Tissues, SSResponse.SSPvalues);
            });
        });
    }

    function transposeTable() {
        transpose = !transpose;
        d3.json("{{ url_for('static', filename=SSPvalues_file) }}").then(SSResponse => {
            buildTable(SSResponse.Genes, SSResponse.Tissues, SSResponse.SSPvalues, transpose);
        });
    }

    init();

    // If unchecked p-value filter, redraw all points:
    function plot_fullgwas(pval_filter_box) {
        d3.select('#plot').text("");
        d3.json("{{ url_for('static', filename=sessionfile) }}").then(response => {
            d3.json("{{ url_for('static', filename=genesfile) }}").then(genesResponse => {
                var data = response;
                var genesdata = genesResponse;
                if(pval_filter_box.checked === false) {
                    // d3.select('#plot_message').append("p").attr("class","text-warning").text(`Please wait while re-drawing with all GWAS points (slow)`);
                    d3.select('#plot_message').append("i").classed('fa',true).classed('fa-info-circle', true).text(`Please wait while re-drawing with all GWAS points (slow)`);
                    plot_gwas(data, genesdata, 
                    eqtl_smoothing_window_size = -1,
                    percent_occupied_by_one_char = 0.020,
                    inputHeight = 720,
                    inputWidth = 1080,
                    font_size = 14,
                    legend_offset = 0.1,
                    pval_filter = false);
                    d3.select('#plot_message').text("");
                }
                else {
                    // d3.select('#plot_message').append("p").attr("class","text-warning").text(`Please wait while re-drawing`);
                    d3.select('#plot_message').append("i").classed('fa',true).classed('fa-info-circle', true).text(`Please wait while re-drawing`);
                    plot_gwas(data, genesdata);
                    d3.select('#plot_message').text("");
                }
            });
        });
    }

    // Listen for redraw click events:
    var coloc_plot_redraw_btn = d3.select('#colocPlot-redraw-btn');
    coloc_plot_redraw_btn.on("click", function() {
        d3.event.preventDefault();

        // Get input values:
        var coloc_plot_width = d3.select('#colocPlotWidth').property("value");
        var coloc_plot_height = d3.select('#colocPlotHeight').property("value");
        var eqtl_window_size = d3.select('#eqtlWindow').property("value");
        var percent_occupied_by_one_char = d3.select('#pctOneChar').property('value');
        var coloc_plot_fontsize = d3.select('#colocPlotFontSize').property('value');
        var legendOffset = d3.select('#legendOffset').property('value');

        if(!coloc_plot_width) {
            coloc_plot_width = 1080;
        }
        if(!coloc_plot_height) {
            coloc_plot_height = 720;
        }
        if(!eqtl_window_size) {
            eqtl_window_size = -1;
        }
        if(!percent_occupied_by_one_char) {
            percent_occupied_by_one_char = 0.1;
        }
        if(!coloc_plot_fontsize) {
            coloc_plot_fontsize = 14;
        }
        if(!legendOffset) {
            legendOffset = 0.1;
        }

        // Clear current colocalization plot
        d3.select('#plot').text("");

        // Redraw colocalization plot with updated parameters
        d3.json("{{ url_for('static', filename=sessionfile) }}").then(response => {
            d3.json("{{ url_for('static', filename=genesfile) }}").then(genesResponse => {
                var data = response;
                var genesdata = genesResponse;
                plot_gwas(data, genesdata, 
                    eqtl_smoothing_window_size = +eqtl_window_size, 
                    percent_occupied_by_one_char = +percent_occupied_by_one_char,
                    inputHeight = +coloc_plot_height, 
                    inputWidth = +coloc_plot_width, 
                    font_size = +coloc_plot_fontsize, 
                    legend_offset = +legendOffset);
            });
        });

        // initialize popover and tooltip
        $(function () {
            $('[data-toggle="popover"]').popover()
        });
        $(document).ready(function(){
            $('[data-toggle="tooltip"]').tooltip({
                delay: { "show": 500, "hide": 100 }
            });
        });
    });

    // Listen for svg plot requests for colocalization plot:
    var coloc_plot_svg_btn = d3.select('#colocPlot-svg-btn');
    coloc_plot_svg_btn.on("click", function() {
        d3.event.preventDefault();
        
        colocPlot_container = d3.select('#plot')
            .select('div.plot-container')
            .select('div.svg-container')
            ;
        
        // Get input values:
        var coloc_plot_width = d3.select('#colocPlotWidth').property("value");
        var coloc_plot_height = d3.select('#colocPlotHeight').property("value");
        
        if(!coloc_plot_width) {
            coloc_plot_width = 1080;
        }
        if(!coloc_plot_height) {
            coloc_plot_height = 720;
        }
        
        saveSvg(colocPlot_container, 'colocalization-plot.svg', coloc_plot_width, coloc_plot_height);
    });

    function saveSvg(svgDiv, name, w = 1080, h = 720) {
        var svgEls = svgDiv.selectAll('.main-svg');
        var head = `<svg class="main-svg" xmlns="http://www.w3.org/2000/svg" xmlns:xlink="http://www.w3.org/1999/xlink" width="${w}" height="${h}" style="background: rgb(255, 255, 255);">`;

        var svgData = head;
        svgEls.each(function(d,i) {
            svgEl = d3.select(this);
            svgData += svgEl.html();
        });
        svgData += "</svg>"
        svgData = svgData.replace('"<', '"&lt;');
        svgData = svgData.replace('<br>','\\r\\n');
        // console.log(svgData);
        var preface = '<?xml version="1.0" standalone="no"?>\r\n';
        var svgBlob = new Blob([preface, svgData], {type:"image/svg+xml;charset=utf-8"});
        saveAs(svgBlob, name);
    }





    // Listen for redraw click events:
    var heatmap_redraw_btn = d3.select('#heatmap-redraw-btn');
    heatmap_redraw_btn.on("click", function() {
        d3.event.preventDefault();

        // Get width and height
        var widthinput = +d3.select('#heatmapWidth').property("value");
        var heightinput = +d3.select('#heatmapHeight').property("value");
        var heatmap_fontsize = +d3.select('#heatmapFontsize').property("value");

        if(!widthinput) {
            widthinput = 1080;
        }
        if(!heightinput) {
            heightinput = 720;
        }
        if(!heatmap_fontsize) {
            heatmap_fontsize = 12;
        }

        // Clear current heatmap
        d3.select('#heatmap').text("");

        // Redraw heatmap with updated width and height
        d3.json("{{ url_for('static', filename=SSPvalues_file) }}").then(SSResponse => {
            plot_heatmap(SSResponse.Genes, SSResponse.Tissues, SSResponse.SSPvalues, widthinput, heightinput, heatmap_fontsize);
        })
    });

    // Listen for svg plot requests for heatmap plot:
    var heatmap_svg_btn = d3.select('#heatmap-svg-btn');
    heatmap_svg_btn.on("click", function() {
        d3.event.preventDefault();

        heatmap_container = d3.select('#heatmap')
            .select('div.plot-container')
            .select('div.svg-container')
            ;
        
        // Get input values:
        var widthinput = +d3.select('#heatmapWidth').property("value");
        var heightinput = +d3.select('#heatmapHeight').property("value");
        
        if(!widthinput) {
            widthinput = 1080;
        }
        if(!heightinput) {
            heightinput = 720;
        }
        
        saveSvg(heatmap_container, 'SS-heatmap.svg', widthinput, heightinput);
    });

  </script>
  <!-- JQuery -->
  <script type="text/javascript" src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.3.1/jquery.min.js"></script>
  <script src="{{ url_for('static', filename='js/navigation-bar.js') }}"></script>
  <!-- Bootstrap tooltips -->
  <script type="text/javascript" src="https://cdnjs.cloudflare.com/ajax/libs/popper.js/1.14.4/umd/popper.min.js"></script>
  <!-- Bootstrap core JavaScript -->
  <script type="text/javascript" src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/4.3.1/js/bootstrap.min.js"></script>
  <!-- MDB core JavaScript -->
  <script type="text/javascript" src="https://cdnjs.cloudflare.com/ajax/libs/mdbootstrap/4.7.4/js/mdb.min.js"></script>
  <!-- DataTables -->
  <script type="text/javascript" src="{{ url_for('static', filename='js/jquery.dataTables.min.js') }}"></script>
  <script type="text/javascript" src="{{ url_for('static', filename='DataTables/datatables.min.js') }}"></script>

</body>


</html>
